src/Storages/MergeTree/MergeTreeIndexUSearch.cpp

```diff
@@ -165,12 +177,16 @@ void MergeTreeIndexAggregatorUSearch<Metric>::update(const Block & block, size_t
         if (!index->reserve(unum::usearch::ceil2(index->size() + num_rows)))
             throw Exception(ErrorCodes::CANNOT_ALLOCATE_MEMORY, "Could not reserve memory for usearch index");
 
-        if (auto rc = index->add(index->size(), array.data()); !rc)
-            throw Exception(ErrorCodes::INCORRECT_DATA, rc.error.release());
-        for (size_t current_row = 1; current_row < num_rows; ++current_row)
-            if (auto rc = index->add(index->size(), &array[offsets[current_row - 1]]); !rc)
+        for (size_t current_row = 0; current_row < num_rows; ++current_row)
+        {
+            auto rc = index->add(static_cast<uint32_t>(index->size()), &array[offsets[current_row - 1]]);
```

Comment:
If I see correctly, one of these overloads is called:
```cpp
add_result_t add(key_t key, b1x8_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_b1x8); }
add_result_t add(key_t key, i8_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_i8); }
add_result_t add(key_t key, f16_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_f16); }
add_result_t add(key_t key, f32_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_f32); }
add_result_t add(key_t key, f64_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_f64); }
```
It looks like we need to convert each input value from single-precision float to half-precision float first (something like [this](https://stackoverflow.com/questions/1659440/32-bit-to-16-bit-floating-point-conversion))?

Reply:
Oh, it seems they do the casting inside `add()` by themselves.

Reply:
Yes, we do. Type-punning is one of the core purposes of the `index_dense_gt` layer.